mesonbuild/compilers/d.py

```diff
@@ -0,0 +1,318 @@
+            if mode == 'link':
+                # Add link flags needed to find dependencies
+                args += d.get_link_args()
+
+        if mode == 'compile':
+            # Add DFLAGS from the env
+            args += env.coredata.external_args[self.language]
+        elif mode == 'link':
+            # Add LDFLAGS from the env
+            args += env.coredata.external_link_args[self.language]
+        # extra_args must override all other arguments, so we add them last
+        args += extra_args
+        return args
+
+    def compiles(self, code, env, extra_args=None, dependencies=None, mode='compile'):
+        args = self._get_compiler_check_args(env, extra_args, dependencies, mode)
+
+        with self.compile(code, args, mode) as p:
+            return p.returncode == 0
+
+    def has_multi_arguments(self, args, env):
+        return self.compiles('int i;\n', env, extra_args=args)
+
+    @classmethod
+    def translate_args_to_nongnu(cls, args):
+        dcargs = []
+        # Translate common arguments to flags the LDC/DMD compilers
+        # can understand.
+        # The flags might have been added by pkg-config files,
+        # and are therefore out of the user's control.
+        for arg in args:
+            if arg == '-pthread':
+                continue
+            if arg.startswith('-Wl,'):
+                linkargs = arg[arg.index(',') + 1:].split(',')
+                for la in linkargs:
+                    dcargs.append('-L' + la.strip())
+                continue
+            elif arg.startswith('-l'):
+                # translate library link flag
+                dcargs.append('-L' + arg)
+                continue
+            elif arg.startswith('-L/') or arg.startswith('-L./'):
+                # we need to handle cases where -L is set by e.g. a pkg-config
+                # setting to select a linker search path. We can however not
+                # unconditionally prefix '-L' with '-L' because the user might
+                # have set this flag too to do what it is intended to for this
+                # compiler (pass flag through to the linker)
+                # Hence, we guess here whether the flag was intended to pass
+                # a linker search path.
+                dcargs.append('-L' + arg)
+                continue
+            dcargs.append(arg)
+
+        return dcargs
+
+
+class GnuDCompiler(DCompiler):
+    def __init__(self, exelist, version, is_cross):
+        DCompiler.__init__(self, exelist, version, is_cross)
+        self.id = 'gcc'
+        default_warn_args = ['-Wall', '-Wdeprecated']
+        self.warn_args = {'1': default_warn_args,
+                          '2': default_warn_args + ['-Wextra'],
+                          '3': default_warn_args + ['-Wextra', '-Wpedantic']}
+        self.base_options = ['b_colorout', 'b_sanitize', 'b_staticpic']
+
+    def get_colorout_args(self, colortype):
+        if mesonlib.version_compare(self.version, '>=4.9.0'):
```

Comment:
__[flake8]__
* [F405] 'mesonlib' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->